<a href="https://colab.research.google.com/github/MarioVidal-Co/Analitica/blob/main/Reto_1_Mario_Vidal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

1. Se cargan los datos

In [ ]:
nxl= '/content/1. BD1_Longitudinal Survey.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB = XDB.dropna() # Assign the result of dropna back to the DataFrame
XDB=XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban']]
XD=np.array(XDB[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize']])
yd=np.array(XDB['Urban'])
XDB=np.array(XDB)
print(XDB)

mnd=GaussianNB()
mnd.fit(XD,yd)
ydp=mnd.predict(XD)

1-np.sum(yd)/len(yd)

2. Implementamos el Modelo

In [ ]:
mnd=GaussianNB()
mnd.fit(XD,yd)
ydp=mnd.predict(XD)

In [ ]:
u=mnd.theta_
sigma=np.sqrt(mnd.var_)

In [ ]:
LI=u-sigma
LS=u+sigma
print("La media de las variables por categoría")
print(u)
print("El limite inferior de la variación de las variables por categoría es")
print(LI)
print("El limite superior de la variación de las variables por categoría es")
print(LS)

La media de las variables por categoría
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
El limite inferior de la variación de las variables por categoría es
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
El limite superior de la variación de las variables por categoría es
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.45637416e+01 1.53621047e+01 7.11164151e+01
  1.74852297e+02 6.95511211e+04 2.67854344e+01 7.12564594e+00]]


In [ ]:
cm=confusion_matrix(yd,ydp)
print(cm)

[[   2 1171]
 [   4 4506]]


In [ ]:
#Intervalo de Variación de la Edad
LI_age = LI[0, 0]
LS_age = LS[0, 0]

print(f"El Intervalo de Variación de la Edad es: [{LI_age:.2f}, {LS_age:.2f}]")

El Intervalo de Variación de la Edad es: [15.90, 20.86]


In [ ]:
LI_height = LI[0, 3]
LS_height = LS[0, 3]

print(f"El Intervalo de Variación de la Altura es: [{LI_height:.2f}, {LS_height:.2f}]")

El Intervalo de Variación de la Altura es: [63.14, 71.29]


In [ ]:
LI_weight = LI[0, 4]
LS_weight = LS[0, 4]

print(f"El Intervalo de Variación del Peso de las Personas es: [{LI_weight:.2f}, {LS_weight:.2f}]")

El Intervalo de Variación del Peso de las Personas es: [118.75, 178.97]


3. Se procede a la evaluación del Modelo

In [ ]:
from re import S
VN=cm[0,0] #Verdaderos Negativos
FP=cm[0,1] #Falsos Positivos
FN=cm[1,0] #Falsos Negativos
VP=cm[1,1] #Verdadeos Positivos

Exactitud=(VP+VN)/len(XD)
print("La exactitud es:",Exactitud) #Comportamiento general

TE=(FN+FP)/len(XD)
print("El Tasa de Error es:",TE) #Porcentaje equivocaciones

Sensibilidad=VP/(VP+FP)
print("La sensibilidad es:",Sensibilidad) #Como se comportó frente a los positivos

Especificidad=VN/(VN+FN)
print("La especificidad es:",Especificidad) #Como se comportó frente a los negativos

La exactitud es: 0.7932430054548654
El Tasa de Error es: 0.2067569945451346
La sensibilidad es: 0.793729082261758
La especificidad es: 0.3333333333333333


4. ¿Qué pasa con una sola persona?

In [ ]:
Xun=np.array([[15,12,12,65,131,25000,26,3]])  #Misma persona del excel
ydun=mnd.predict(Xun)
print(ydun)

[1.]


**Respuestas**

**1.** El número total de la base de datos son 12678

**2.** El porcentaje de datos que hacen parte de las categorías de clasificación es de 79.35% y los que no hacen parte es de 20.64%

**3.** El intervalo de variación de la edad es [15.90, 20.86]

**4.** El intervalo de variación de la Altura es: [63.14, 71.29]

**5.** El intervalo de variación del Peso de las Personas es: [118.75, 178.97]

**6.** La exactitud del modelo es de 79.32%

**7.** La tasa de error es del 20.68%

**8.** La sensibilidad es de 79.37%

**9.** La Especificidad es de 33.33%

**10.** Persona con las siguientes características indicar si vive en la ciudad o las áreas rurales 15,12,12,65,131,25000,26,3 (edad, educación de madre, educación de padre, altura, peso, ingresos, autoestima, tamaño de la familia). Esta persona vive en el Área Urbana.

**Análisis**

Con este modelo podemos concluir que, la mayoría de personas en la base de datos residen en el Área Urbana ya que notamos que el tamaño de la familia y el ingreso es mayor en esta zona.

De acuerdo a las métricas este modelo obtuvo una certeza de 79.37%, lo que nos permite afirmar que este modelo si permite determinar la zona de residencia de una persona, además presenta un 79.32% de exactitud al momento de evaluar las variables, de igual forma arrojó una tasa de error del 20.67%.

Por último, la persona escogida para evaluar el modelo arrojó un resultado positivo [1] para residente en área urbana y comparando este resultado con la hoja de MS Excel se pudo re-afirmar el resultado y categorizar el modelo como exitoso.
